src/datatype.c

```diff
@@ -25,6 +25,39 @@ static int is10digit(char c) JL_NOTSAFEPOINT
     return (c >= '0' && c <= '9');
 }
 
+static int is_canonicalized_anonfn_typename(char *name) JL_NOTSAFEPOINT
+{
+    assert(strlen(name) > 1);
+    // first character after '#' is a digit: canonicalized anonymous function name
+    if (is10digit(name[1])) {
+        return 1;
+    }
+    char *outermost_func_name = NULL;
+    // go backwards from name + strlen(name) to find the first `_`
+    for (char *c = name + strlen(name); c > name; c--) {
+        if (*c == '_') {
```

Comment:
Does this still work for names like `do_foo`?

If so, it would be good to explain in the comments why it's OK.